README.md

```diff
@@ -29,122 +37,1058 @@ The library design is inspired by the work in [Exploring HTN Planners through ex
+    classDef selected stroke:#0f0
+```
+
+In the diagram, the nodes connected via the dotted lines show the planning process, while nodes connected via the solid arrows show the resulting plan. The diagram shows how the planner tries first the `+2` method, which results into the expansion `('+1', '+1')`, a second tries fails, because the condition no longer holds, so the planner adds a `+1` action next, which allows the planner to reach the target. If we compare this diagram with one where the planner doesn't know have the `plusTwo` method, we get.
+
+```mermaid
+graph TD
+    start(( ))
+    start -.- d0{ }
+    d0 -.- 2586e9e("+1")
+    2586e9e -.- d1{ }
+    d1 -.- 82f6258("+1")
+    82f6258 -.- d2{ }
+    d2 -.- a1f7280("+1")
+    a1f7280 -.- stop(( ))
+    stop:::finish
+    classDef finish stroke:#000,fill:#000
+    start:::selected
+    start --> 2586e9e
+    2586e9e:::selected
+    2586e9e --> 82f6258
+    82f6258:::selected
+    82f6258 --> a1f7280
+    a1f7280:::selected
+    a1f7280 --> stop
+    classDef error stroke:#f00
+    classDef selected stroke:#0f0
+```
+
+Here we see here that the planner needs to perform 3 iterations (the diamond nodes) to find a plan to the target.
+
+Methods can also call other methods. We can now write `plusThree` using `plusOne` and `plusTwo`
+
+```typescript
+const plusThree = Task.from<number>({
+	condition: (state, { target }) => target - state > 2,
+	// Methods can be referenced from other methods
+	method: (_, { target }) => [plusTwo({ target }), plusOne({ target })],
+	description: '+3',
+});
+```
+
+Methods, however cannot call themselves recursively or as part of call chain. This is forbidden to prevent infinite recursion. Using recursion means the method will effectively never be called.
+
+## Lenses
+
+Let's say now that we want our agent to manage multiple counters. Let's redefine our system state once more to do this
+
+```typescript
+type System = { counters: { [key: string]: number } };
+```
+
+We can now easily (?) redefine our `plusOne` task to handle this case.
+
+```typescript
+const plusOne = Task.from<System>({
+	// This task will be chosen only if one of the keys is smaller than the target
+	condition: (state, { target }) =>
+		Object.keys(state.counters).some(
+			(k) => state.counters[k] < target.counters[k],
+		),
+	effect: (state, { target }) => {
+		// We find the first counter below the target, we know it exists because of
+		// the condition so we use the non-null assertion (!) at the end
+		const key = Object.keys(state._.counters).find(
+			(k) => state._.counters[k] < target.counters[k],
+		)!;
+
+		// Update the changed counter
+		state._.counters[key]++;
+	},
+	description: '+1',
+});
+```
+
+The above code achieves the same as the single counter `plusOne` (increase one of the counters), but is a bit convoluted, and we cannot tell from the description which counter will be increased. We can improve this setting the `lens` property when defining the action.
```

Comment:
```suggestion
The above code achieves the same as the single counter `plusOne` (increase one of the counters), but is a bit convoluted, and we cannot tell from the description which counter will be increased. We can improve this by setting the `lens` property when defining the action.
```

My brain parsed the text as "We can improve this by setting". The "by" helps to direct the brain parser to the right path without backtracking :-)